### ISSUE TITLE
Make uber-contracts page for ALL APIs across all courses

### ISSUE THREAD

**Issue by schanzer** [bootstrapworld/curriculum#1453]
- [x] Make this page
- [x] Get rid of @pathway-link
- [x] Add blank DR (WeScheme and Pyret!), put in static pages directory. Link from everywhere!
- [x] Add blank DC, put in static pages directory

**Comment by schanzer** [comment]
@flannery-denny the `contracts-dictionary` branch now includes a page called Contracts.shtml that lives at the top level of the distribution directory, alongside Alignments.shtml and Glossary.shtml.

I made the language table JSON in a hurry just for a proof of concept, so it's incomplete and likely contains a few errors.

Is this what you had in mind?

**Comment by flannery-denny** [comment]
Thanks @schanzer!
I'm noticing:
- text at top should be different for this page
- all of the contracts I'm seeing here are simple form and many will need to be here in their annotated form
- all contracts in the web version should have examples
-  I wonder if this is an opportunity to include purpose statements?

**Comment by schanzer** [comment]
@flannery-denny pushed a change that adjusts the text at the top. Please feel free to tweak, or let me know what you want it to say.

Added examples for everything. It would really help me out if you could add some annotations to the langTable, but if not I will get to it later. 

Purpose statements! I love this idea, but I _really_ love it if we can replace the examples with them and force a little more reading comprehension.

I've also added an example of how to deal with links (see lessons/contracts/index.adoc). If this works for you, it's easy to find and replace this prose across all other lessons. 

Happy to add a link to the TR section once the rest is done.

**Comment by flannery-denny** [comment]
@schanzer 
- I'd love to help you out with the lang table work, but I don't understand where to find it or edit it.
- left you a comment on the link commit with a screenshot - didn't work for me
- and even with make force and force reload I'm not able to see the updates to the uber contracts. are you sure you pushed to the contracts-dictionary branch?

**Comment by schanzer** [comment]
@flannery-denny the table is defined in `lib/langtable.js`. (You can always look at the commits for this branch to find out which files I've been messing with).

Yes, the updated files have all been pushed. Are you sure you've checked out the right branch and done `git pull`?

**Comment by flannery-denny** [comment]
@schanzer This looks better! And convinces me that it would be way more meaningful as a reference if it included purpose statements... but I'm thinking that having access to that resource might sometimes undermine lesson structures. Noticing some discrepancies between the function titles in contracts and examples 
<img width="467" alt="Screen Shot 2023-08-09 at 6 03 11 PM" src="https://github.com/bootstrapworld/curriculum/assets/66691020/192e79f2-4b7b-4af7-a44d-51eccad69528">

**Comment by flannery-denny** [comment]
@schanzer Happy to complete the annotations for the language table and can work on it tonight or tomorrow afternoon once I understand the correct formatting. Not seeing an example of that.

Currently we have entries like:

```
{"name": "rectangle",
       "domain": "(Number, Number, String, String)",
       "range": "Image",
       "example": "rectangle(100, 50, \"outline\", \"green\")"},
```

Should annotations look like this?

```
{"name": "rectangle",
       "domain": "((width Number), (length Number), (fill-style String), (color String))",
       "range": "Image",
       "example": "rectangle(100, 50, \"outline\", \"green\")"},
```

**Comment by flannery-denny** [comment]
@schanzer I've been thinking about the way this uber-contracts page is getting built and am curious why you wouldn't want to take advantage of @ds26gte 's `@contract` work? 

Either way, letting you know that the work on this that I was expecting from you was to figure out how to link to it. I thought this page would just be generated by copying all of the contracts from algebra into a document and adding in the DS contracts that don't overlap.

Also want to flag that since the existing alg and DS contracts pages use the old expanded form of contracts, and aren't using the contracts directive, they are going to have to get generated in a new way in order to incorporate the annotated formatting...

| `pie-chart-summarized`
| `{two-colons}`
| `(t {two-colons} Table, labels {two-colons} String, values {two-colons} String)`
| `->`
| `Image`
5+|`_pie-chart-summarized(animals-table, "age", "pounds")_`

**Comment by flannery-denny** [comment]
@schanzer I've removed the link you added from contracts, because the lesson doesn't work as written if we give them a contracts page with all of the contracts filled in. Instead, I've added a copy of your link to `making-game-images`

**Comment by flannery-denny** [comment]
@schanzer I was very excited to read that you'd implemented @ds26gte 's fitb work on contracts, but I've pulled and used make force on the `contracts-dictionary` branch and I am seeing the old version of expanded contracts instead of the new annotated contracts in both the pdf accessible from DS resources and the ueber contracts file linked from making game images.  Am I missing something?

<img width="891" alt="Screen Shot 2023-08-10 at 3 13 17 PM" src="https://github.com/bootstrapworld/curriculum/assets/66691020/441c921e-25f6-46cb-863e-9aa40fa8c9fa">

Looks like you went ahead and did all the grunt work of writing out annotated contracts rather than writing me a spec, but if there's something left for me to do to support you, please lmk.

ps. it would help me in my review of the work if you could tag this issue in your commits.

**Comment by schanzer** [comment]
@flannery-denny I have not implemented any of his work on contracts yet. Will let you know when I do. (#1453)

**Comment by schanzer** [comment]
@flannery-denny DONE. Enjoy 😉

**Comment by schanzer** [comment]
@flannery-denny this table now works with WeScheme, AND the ruby contracts.

**Comment by flannery-denny** [comment]
@schanzer very cool. Thanks!

- Noticing that the examples in this contracts table can be copy and pasted into pyret. Tested for typos and was able to fix a bunch :)
- Can't easily test some of the DS ones like `build-column(animals-table, kilograms)` because they are dependent on something being predefined. If you have a chance to scan those contracts and examples, perhaps you'll catch something I missed?
- Would it be possible to not require annotations in the language table? There are a few contracts that would be easier to read without annotation and currently if I leave them off this is what I get:
<img width="93" alt="Screen Shot 2023-08-10 at 5 57 42 PM" src="https://github.com/bootstrapworld/curriculum/assets/66691020/55b28f9d-a79e-40f0-83bc-c63e94f40f75">

- Can we deploy the strategy you employed here for the contracts pages at the back of the book? They aren't utilizing @fitb yet... and they need to be building in the same style before we can migrate these annotated contracts to master.
- Dorai told me that his ruby code checked whether the Type or Annotation had more characters and decided the length of the ruby based on the greater one, but I'm seeing this issue where clearly the length is being tested by the Type and not the annotation... and there is definitely room for the annotation here.
<img width="799" alt="Screen Shot 2023-08-10 at 6 14 20 PM" src="https://github.com/bootstrapworld/curriculum/assets/66691020/c7798027-b485-49da-93e9-17c6eecb0f91">

**Comment by schanzer** [comment]
@flannery-denny 
* I looked at the other examples that use helper functions, and aside from the one you listed they seem right to me. I've pushed a fix.
* I've pushed a change that will drop the ruby if the `name` is missing from langTable, OR if it's there but it's the empty string.
* Porting this code to the contracts at the end of the book is not a good idea. The way to go from here is to REPLACE those pages with _this_. That will require a little bit of collaboration between Dorai and I, but stay tuned for good things.

**Comment by flannery-denny** [comment]
Need to add the following contracts from DS:

- [ ] `first-n-rows`
- [ ] `line-graph`
- [ ] `num-max`
- [ ] `num-min`
- [ ] `r-value`
- [ ] `sum`

**Comment by flannery-denny** [comment]
@schanzer 
- The minimum size for the ruby looks fine to me.
- I can live with it as is but 
-- if it's easy to put a fitb without a ruby under the types that don't get annotations that would increase consistency of appearance.
-- colors feel inconsistent with our other stuff, but I imagine it's an easy fix and they'll all change soon anyway
- excited to see what you and Dorai come up with for the end of the book. 
-- Reminding you that last we talked about it, the contracts in the back of the Algebra book were not supposed to have example code filled in.

**Comment by schanzer** [comment]
@flannery-denny 
* added fitb when a ruby _just won't do_
* what colors don't feel consistent? we are using the same css rules for color
* re:missing contracts for algebra -- I can do that if it's _all_ contracts. You OK leaving _every_ contract blank for Algebra?

**Comment by flannery-denny** [comment]
re: hiding algebra examples in contracts - I'm totally fine with that! we currently show 5 examples (`num-sqr`, `num-sqrt`, `string-contains`, `string-length`, and `triangle`) and I think that's a bug.

**Comment by flannery-denny** [comment]
Re: colors... I guess it's just the age-old website struggle with, "what does purple look good with, anyway?" _(sidenote: I'm not totally anti purple! siberian irises and louisiana irises are about as gorgeous as it gets!)_ and we should ignore it for now... but this big swath of purple at the top followed by the contracts table in a completely other color scheme feels jarring to me.
<img width="846" alt="Screen Shot 2023-08-11 at 7 28 36 AM" src="https://github.com/bootstrapworld/curriculum/assets/66691020/2689e042-4afc-432b-83cd-b1a9b6e7caf3">

**Comment by schanzer** [comment]
@flannery-denny phew! For a moment there I thought I was going colorblind. :)

Just pushed a fix that disables domains and examples for our algebra pathways.

**Comment by flannery-denny** [comment]
@schanzer There's something strange about the current positioning of the contracts table. I have my browser at the full width of my screen. 

This is what the contracts table looks like - it has almost no right margin! (I know it's a landscape page... but it also starts farther to the right than our regular pages...) 

See second image from a lesson  plan and note how the left margin compares to the exclamation point.
<img width="1459" alt="Screen Shot 2023-08-11 at 8 29 31 AM" src="https://github.com/bootstrapworld/curriculum/assets/66691020/0438ec25-491c-4653-b117-e0aa383a98db">

<img width="1466" alt="Screen Shot 2023-08-11 at 8 29 53 AM" src="https://github.com/bootstrapworld/curriculum/assets/66691020/86e617d7-4ec1-4970-bdc5-2d2f6e5bcf42">

**Comment by flannery-denny** [comment]
Also, the last time I looked at this page, all of the rubys fit, and the page was not this wide.

**Comment by flannery-denny** [comment]
And, in the current iteration, the title for Domain is to the right of all of the content for Domain.

**Comment by schanzer** [comment]
@flannery-denny nothing strange about it! That's how the pathway pages look as well. That space is filled by the Bootstrap menu bar on the web site. The lesson plans start just as far out to the right as well. I think you're only noticing it here because it doesn't have the lesson toolbar, and it's been a while since you really looked at the pathway pages.

Sounds like you don't want the column headings to be centered, so I'll push them back to the left.

**Comment by schanzer** [comment]
Closing this since a separate issue exists for the PR, which is pending review
